pcap-ioplugin.c

```diff
@@ -0,0 +1,259 @@
+#include <sys/bitypes.h>
+#endif
+#include <sys/types.h>
+#endif /* _WIN32 */
+
+#include <errno.h>
+#include <memory.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "pcap-int.h"
+
+#ifdef HAVE_DLFCN_H
+#include <dlfcn.h>
+#endif
+
+#ifdef HAVE_OS_PROTO_H
+#include "os-proto.h"
+#endif
+
+static FILE*
+stdio_open_read(const char *fname, char *errbuf)
+{
+	FILE *fp = NULL;
+
+	if (fname == NULL) {
+		snprintf(errbuf, PCAP_ERRBUF_SIZE,
+		    "A null pointer was supplied as the file name");
+		return (NULL);
+	}
+
+        if (fname[0] == '-' && fname[1] == '\0')
+        {
+                fp = stdin;
+                if (!stdin || ferror(stdin)) {
+                        snprintf(errbuf, PCAP_ERRBUF_SIZE,
+                                 "The standard input is not open");
+                        return (NULL);
+                }
+		SET_BINMODE(fp);
+	} else {
+		/*
+		 * "b" is supported as of C90, so *all* UN*Xes should
+		 * support it, even though it does nothing.  It's
+		 * required on Windows, as the file is a binary file
+		 * and must be written in binary mode.
+                 *
+                 * Use charset_fopen(); on Windows, it tests whether we're
+                 * in "local code page" or "UTF-8" mode, and treats the
+                 * pathname appropriately, and on other platforms, it just
+                 * wraps fopen().
+                 *
+                 */
+                fp = charset_fopen(fname, "rb");
+		if (fp == NULL) {
+			snprintf(errbuf, PCAP_ERRBUF_SIZE, "%s: fopen: %s", fname,
+				pcap_strerror(errno));
+			return (NULL);
+		}
+	}
+
+	return fp;
+}
+
+static FILE*
+stdio_open_write(const char *fname, char *errbuf)
+{
+	FILE *fp = NULL;
+
+	if (strcmp(fname, "-") == 0) {
+		fp = stdout;
+                if (!stdout || ferror(stdout)) {
+                        snprintf(errbuf, PCAP_ERRBUF_SIZE,
+                                 "The standard input is not open");
+                        return (NULL);
+                }
+
+		SET_BINMODE(fp);
+	} else {
+		/*
+		 * "b" is supported as of C90, so *all* UN*Xes should
+		 * support it, even though it does nothing.  It's
+		 * required on Windows, as the file is a binary file
+		 * and must be written in binary mode.
+		 */
+                fp = charset_fopen(fname, "rb");
+		if (fp == NULL) {
+			snprintf(errbuf, PCAP_ERRBUF_SIZE, "%s: fopen: %s", fname,
+				pcap_strerror(errno));
+			return (NULL);
+		}
+	}
+
+	return fp;
+}
+
+static const pcap_ioplugin_t*
+pcap_ioplugin_stdio(void) {
+	static pcap_ioplugin_t plugin = {
+		.open_read = stdio_open_read,
+		.open_write = stdio_open_write
+	};
+
+	return &plugin;
+}
+
+/*
+ * loads an I/O plugin with the given name
+ * (on UNIX, a .so shared library)
+ *
+ * NB: fails silently and falls back to existing uncompressed
+ *     stdio-based output if the plugin fails to load
+ */
+
+const pcap_ioplugin_t*
+pcap_ioplugin_init(const char *name)
+{
+	void *lib = NULL;
```

Comment:
```
./pcap-ioplugin.c: In function 'pcap_ioplugin_init':
./pcap-ioplugin.c:159:8: warning: unused variable 'lib' [-Wunused-variable]
  159 |  void *lib = NULL;
      |        ^~~
```

Reply:
There are several issus here.

1. On UN*X, we need either to assume we have the `dlopen()` APIs or check for it in CMakeLists.txt as well.  It's specified in POSIX and, as far as I know, the only general-purpose UN*Xes that don't have it we're likely to encounter are 1) 32-bit HP-UX (which has its own API) and 2) perhaps older versions of AIX.  However, perhaps some embedded Linux distributions have a cut-back libc that doesn't have it.
2. We're not handling Windows which has its own APIs for this.
3. If we don't have any way to load plugins, we shouldn't define a variable to hold the plugin handle.